docs/proposals/action_based_roles.md

```diff
@@ -0,0 +1,156 @@
+underlying tasks, variables and other Ansible resources are shared. E.g. templates
+of service files, handlers, sequence of installation or configuration tasks.
+
+Additionally, some operations can be unified into one and the proper behavior hidden
+inside a role as implementation details. E.g. component upgrade can be equivalent
+to component installation up to some differences. The role dealing with the installation
+should then decide if the actual installation or rather an update is required.
+
+In practice, it is natural to create a role that deals with an installation as
+the first step. Later, role that deals with an upgrade is implemented. Not rarely
+followed by implementation of other operations. At the end one can end up
+with a set of roles, each responsible for a single operation. Naturally,
+the layout leads to code duplication and various efforts to minimize it.
+
+## Design
+In case a set of roles conforms to the following conditions:
+
+* each role is responsible for a distinct operation (e.g. installation, upgrade, migration)
+* all operations are applied on the same component
+* the operations share common code base and Ansible resources
+
+consolidate all roles into new one, providing a mechanism to select specific
+operation to apply when the new role is invoked. Since Ansible 2.2 one can
+use `include_role` module and perform file-based operation selection:
+
+```yaml
+include_role:
+  name: <role>
+tasks_from: <operation>
+```
+
+Additionally, one can define a minimal set of actions that each action-based
+role should provide:
+* pre-checks: to check if the role has all necessary conditions fulfilled
+* post-checks: to check a role ended successfully
+
+### Layout
+
+Proposed layout of an action-based role:
+
+```sh
+$ ls <role>/tasks
+<action1_dir>
+<action2_dir>
+...
+<action1>.yml
+<action2>.yml
+...
+```
+
+The role `tasks` directory consists of:
+* one directory per each action (action directory)
+* one yml file per each action that includes a file from corresponding action directory
+* additional yml files needed by the role (e.g. task files shared across actions)
+
+Templates and other resources location are not affected by the layout.
+If reasonable, the action directory layout can be duplicated inside the `templates` directory.
+
+**Expected content of an action file**:
+
+```yaml
+# content of tasks/<action>.yml
+
+# include a main file from action directory
+- name: Running <action>
+  include: <action>/main.yml
+```
+
+If an action directory was created from a previously existing role,
+it contains all the role's tasks files.
+If it was created as a new action, it contains all tasks files needed
+to implement the action.
+
+### Usage
+
+```yaml
+- hosts: <hosts>
+  tasks:
+  ...
+  - include_role:
+      name: <role>
+      tasks_from: <action>
+    vars:
+      <var1>: "{{ <value1> }}"
+      <var2>: "{{ <value2> }}"
+      ...
+```
+
+If the `tasks/main.yml` is present, `include_role` without the `tasks_from`
```

Comment:
I would prefer that main.yml has contents like the following:
https://github.com/mgugino-upstream-stage/openshift-ansible/blob/86dfcb45e4c1149010d1d460d6ca6e0809063e93/roles/openshift_hosted/tasks/main.yml

It may be confusing to users who are browsing the source of the role to try to figure out what the role does.  They'll look at main.yml, see some tasks, and go "What is all this other stuff, how is it being used?"

Reply:
Good point.

Reply:
Updated